### PR TITLE
Fix ASTextNode size calculation when paragraphstyle has linespacing

### LINKED ISSUE #930: (ASDK 1.9.3) ASTextNode truncates on first line instead of second line for single words that are long enough
For "hsiung@gmail.com" (see the last icon and caption in the screenshot below) it renders the text in two lines without truncation. However a longer email address such as "hsiung@mailinator.com" will result in truncation on the first line (see the first icon and caption in the screenshot below). In 1.9.2 ASTextNode was truncating correctly on the second line.

simulator screen shot dec 9 2015 11 09 16 am

Text node:
```
ASTextNode *textNode = NSMutableParagraphStyle *paragraphStyle = [[NSMutableParagraphStyle alloc] init];
        paragraphStyle.lineBreakMode = NSLineBreakByWordWrapping;
        paragraphStyle.paragraphSpacing = paragraphStyle.lineSpacing;
        paragraphStyle.alignment = NSTextAlignmentCenter;
        UIFont *font = [UIFont fontWithName:@"HelveticaNeue-Bold" size:12.0];
        NSMutableDictionary *attributes = [[NSMutableDictionary alloc] initWithDictionary:@{ NSParagraphStyleAttributeName : paragraphStyle,
                                                                                             NSFontAttributeName : font,
                                                                                             NSForegroundColorAttributeName : [UIColor blackColor],
                                                                                             NSKernAttributeName : @(-0.2)}];
        [textNode setAttributedString:[[NSAttributedString alloc] initWithString:@"averylongemail@address.com" attributes:attributes]];
```
Layout code:
```
    CGSize exactSize = CGSizeMake(60, 30);
    ASStackLayoutSpec *verticalStack = [ASStackLayoutSpec verticalStackLayoutSpec];
    verticalStack.alignItems = ASStackLayoutAlignItemsStretch;
    verticalStack.children = @[textNode];
    verticalStack.layoutPosition = CGPointZero;
    verticalStack.sizeRange = ASRelativeSizeRangeMakeWithExactCGSize(exactSize);
    ASStaticLayoutSpec *staticSpec = [[ASStaticLayoutSpec alloc] init];
    staticSpec.children = @[verticalStack];
```

### PR DESCRIPTION
**Pull request by samhsiung**
This fixes #930 and fixes ASTextNodeTest size recalculation tests when there is a non-zero NSParagraphyStyle lineSpacing. 

I could only get the updated ASTextNodeTest tests to pass when I increased the height of the rect returned by usedRectForTextContainer by exactly linespacing in ASTextKitRenderer's `_calculateSize` method. I suspect the calculation doesn't take the linespacing of the last line into account.

I couldn't find any issues within ASDK code so I suspected it is a bug with TextKit or NSLayoutManager but if anyone could find the root cause in ASDK that is obviously a better fix.

### CODE REVIEW

**Comment by bsmith11** [comment]
Could this have anything to do with #940?

**Comment by samhsiung** [comment]
Thanks @bsmith11 that's interesting. Possibly related if linespacing calculation in NSLayoutManager is broken in general, however I wasn't using NSUnderlineStyleAttributeName in my project or in the ASTextNodeTest

**Comment by bsmith11** [comment]
I'll test it out on `UITextView`

File: AsyncDisplayKit/TextKit/ASTextKitRenderer.mm [hunk 1]
```diff
@@ -82,7 +82,20 @@ - (void)_calculateSize
   CGRect constrainedRect = {CGPointZero, _constrainedSize};
   __block CGRect boundingRect;
   [_context performBlockWithLockedTextKitComponents:^(NSLayoutManager *layoutManager, NSTextStorage *textStorage, NSTextContainer *textContainer) {
+    NSParagraphStyle *paragraphStyle = [textStorage attributesAtIndex:[layoutManager characterIndexForGlyphAtIndex:0]
```
Comment (appleguy) [inline AsyncDisplayKit/TextKit/ASTextKitRenderer.mm]: What if the paragraph style changes at some point through the text?  Just not supported?

File: AsyncDisplayKit/TextKit/ASTextKitRenderer.mm [hunk 2]
```diff
@@ -82,7 +82,20 @@ - (void)_calculateSize
...
+                                                       effectiveRange:NULL][NSParagraphStyleAttributeName];
+
     boundingRect = [layoutManager usedRectForTextContainer:textContainer];
+    
+    // Work around line spacing of the last line not getting taken into account
```
Comment (appleguy) [inline AsyncDisplayKit/TextKit/ASTextKitRenderer.mm]: So is this a TextKit bug?  Were you able to find any other reports of it online?  Ideally we should file a Radar for this and reference the number here...almost certainly this code will break in a future version of iOS (if it is a workaround), and thus a future version of ourselves will be puzzling over the intent and context here.

File: AsyncDisplayKit/TextKit/ASTextKitRenderer.mm (hunk 1 above)
Comment (samhsiung) [inline AsyncDisplayKit/TextKit/ASTextKitRenderer.mm]: Good point and you caught my laziness :) modified it to only take the paragraph style of the last visible glyph and it still fixed it.

File: AsyncDisplayKit/TextKit/ASTextKitRenderer.mm (hunk 2 above)
Comment (samhsiung) [inline AsyncDisplayKit/TextKit/ASTextKitRenderer.mm]: Couldn't find anything on google. Logged a bug rdar://23885167.

**Comment by samhsiung** [comment]
closing and reopening to trigger travis build

**Comment by samhsiung** [comment]
After setting up test code to verify if TextKit is bugged, I think the issue might be in the truncation logic.

**Pull request closed without merging**